File: A1/GO/a1.go
package main

import ("fmt"
        )

//combine function takes in two strings x and y and returns the form (x.y)
func combine (x string , y string ) string {
        return "(" + x+ "." + y + ")"

}

//tree enumeration function takes in n as the number of nodes, and outputs a list of all possible binary trees of that number
func treeVal (n int) []string {
        if(n == 0){
                return []string{""}
        }
	//an empty list of strings.
        treeList := []string{}
        for i := 0; i < n; i++ {
                nLeft := i
                nRight := n-i-1
                for _, leftNode  := range treeVal(nLeft){
                        for _, rightNode := range treeVal(nRight) {
				//we append the combination of left and right subtrees into the tree list
                                treeList = append(treeList, combine(leftNode,rightNode))
                        }
                }

        }
        return treeList

}
//function main takes user integer input, does some type checking and then prints the trees each on a new line
func main() {

        var i int
        fmt.Print("Please provide the number of nodes: ")
        _, err := fmt.Scanf("%d", &i);
        if err != nil {
                fmt.Println(err);
        }
        for _, tree := range treeVal(i){
                fmt.Println(tree);
        }
}

File: A1/Python/a1.py
import math
#main function takes in user input, an integer number of nodes, and prints out all the possible binary trees of that number
def main():
   n = eval(input("Please provide the number of nodes: "))
   for tree in treeVal(n):
      print(tree)

#n is the number of nodes
#we return an empty list if the tree has 0 nodes, otherwise we recurse and output every possible combination of left and right subtrees, this algorithim was shared on piazza by Liam Heeger. 
def treeVal(n):
   if(n == 0):
      return [""]
   treeList = []
   for i in range (n):
      nLeft = i
      nRight = n-i-1
      for leftNode in treeVal(nLeft):
         for rightNode in treeVal(nRight):
            treeList.append(combine(leftNode,rightNode))
   return treeList
#combine function takes in two trees (strings) and returns the form (leftTree.rightTree)
def combine(leftTree,rightTree):
   return "(" + leftTree + "." + rightTree + ")"
main()
